enhancements/machine-config/certificate-authorities-for-image-registries.md

```diff
@@ -0,0 +1,185 @@
+  - TBD
+creation-date: 2022-12-01
+last-updated: 2022-12-01
+tracking-link:
+  - https://issues.redhat.com/browse/IR-230
+see-also:
+replaces:
+superseded-by:
+---
+
+# Additional trusted certificate authorities for image registries
+
+## Summary
+
+This enhancement describes how certificate authorities for image registries
+should be distributed to CRI-O, openshift-apiserver, and other clients when the
+image registry operator is not installed.
+
+## Motivation
+
+The image registry operator manages a daemon set that distributes certificate
+authorities for the integrated image registry, and it also distributes
+certificate authorities for external registries that are provided by the
+cluster administrator in the image config. This makes the image registry
+operator required for the cluster while the integrated image registry is
+optional and can be removed.
+
+The goal of this enhancement is to move this required functionality to an
+essential cluster operator, and have the image registry operator only manage
+the optional registry workload.
+
+### User Stories
+
+* As a <role>, I want to <take some action> so that I can <accomplish a
+goal>.
+
+### Goals
+
+* Make additionalTrustedCA be managed by machine-config-operator.
+* Create a mechanism for the cluster-image-registry-operator to distribute
+  certificate authorities for the integrated image registry.
+* Remove the node-ca daemon set.
+
+### Non-Goals
+
+* Change additionalTrustedCA behavior for its users.
+
+## Proposal
+
+Create a controller inside machine-config-operator that will replace the
+node-ca daemon set.
+
+This controller should handle and distribute to all nodes certificate
+authorities from the user-provided config map that is specified in
+`images.config.openshift.io/cluster`. It should also observe a config map in
+the `openshift-config-managed` namespace and merge its content with the
+user-provided config map. This will allow cluster-image-registry-operator to
+provide certificate authorities for the integrated image registry without
+changing the user-provided config map.
+
+Once this controller is created, the node-ca daemon set should be removed.
+
+### Workflow Description
+
+Let's suppose the user creates a config map with the certificate authorities
+for their registries:
+
+```yaml
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: registry-ca
+  namespace: openshift-config
+data:
+  registry.example.com: |
+    -----BEGIN CERTIFICATE-----
+    ...
+    -----END CERTIFICATE-----
+  registry-with-port.example.com..5000:
+    -----BEGIN CERTIFICATE-----
+    ...
+    -----END CERTIFICATE-----
+```
+
+Then updates the `images.config.openshift.io/cluster` object to point to this
+config map:
+
+```yaml
+spec:
+  additionalTrustedCA:
+    name: registry-ca
+```
+
+If the image registry operator is installed, there should be a config map:
```

Comment:
As we are talking here about making image-registry operator optional, what happens when image-registry operator is not installed?

Reply:
The `image-registry-ca` config map is required for the image registry to correctly work in a cluster.
When the image registry operator is not installed in the cluster, the `image-registry-ca` config map will not be present.

The important thing here is that when it is present, the MCO should distribute the CA on the cluster nodes (which the registry operator currently does, but will stop doing).
It's worth noting that this would be the path of least friction. Another option would be to rely on machine configs (created by the registry operator) to distribute the CAs required by the registry, so the MCO would only need a new path for distributing user provided CAs via `additionalTrustedCA` in the `images.config.openshift.io/cluster` (to keep backwards compatibility). The downside of this approach is that could potentially disrupt other openshift components (such as builds) that also rely on the integrated registry CAs.

Reply:
make sense. thanks